### ISSUE TITLE
Add templates field to AlertmanagerConfig CRD

### ISSUE THREAD

**Issue by slyt** [prometheus-operator/prometheus-operator#5139]
**What is missing?**
The `AlertManagerConfig` CRD does not have the ability to add `templates` to the `alertmanager.yaml` config file. It appears the only way to configure the the search paths for the templates is by using a Secret to configure the AlertManager, [as mentioned in the alerting docs](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md#using-a-kubernetes-secret).

**Why do we need it?**
I want to use a notification template in my AlertManagerConfig resources but am unable to do so.

**Environment**

* Prometheus Operator version:

    `Prometheus-Operator v0.60.1`, deployed with kube-prometheus-stack v41.0.0

**Anything else we need to know?**:
I'd be happy to help implement this.

**Comment by JoaoBraveCoding** [comment]
Hello @slyt I think your issue is similar to https://github.com/prometheus-operator/prometheus-operator/issues/4606 which seems to have already been fixed by the PR https://github.com/prometheus-operator/prometheus-operator/pull/4933

**Comment by slyt** [comment]
> Hello @slyt I think your issue is similar to #4606 which seems to have already been fixed by the PR #4933

That looks like it's just for the `alertmanagerConfiguration` within the [Alertmanager CRD](https://github.com/prometheus-operator/prometheus-operator/blob/main/jsonnet/prometheus-operator/alertmanagers-crd.json). 

I'm looking to add template support to the [AlertmanagerConfig CRD](https://github.com/prometheus-operator/prometheus-operator/blob/main/jsonnet/prometheus-operator/alertmanagerconfigs-crd.json).

**Comment by DylanWarnock** [comment]
I've also tried add inline templates:

```
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  labels:
    argocd.argoproj.io/instance: prometheus-alert-manager-config
  name: alertmanager-config-override
  namespace: kube-prometheus-stack
spec:
  receivers:
    - name: prometheus-default
      slackConfigs:
      - sendResolved: true
        apiURL:
          key: prometheusDefaultSlackWebhook
          name: {{ $.Values.encryptedSecrets.name }}
        text: "{{ range .Alerts }}{{ .Annotations.summary }}{{ .Annotations.description }}{{ .Annotations.message }}{{ end }}"
```

but the text field gets applied as an empty string

**Comment by slyt** [comment]
> I've also tried add inline templates:
> 
> ```
> apiVersion: monitoring.coreos.com/v1alpha1
> kind: AlertmanagerConfig
> metadata:
>   labels:
>     argocd.argoproj.io/instance: prometheus-alert-manager-config
>   name: alertmanager-config-override
>   namespace: kube-prometheus-stack
> spec:
>   receivers:
>     - name: prometheus-default
>       slackConfigs:
>       - sendResolved: true
>         apiURL:
>           key: prometheusDefaultSlackWebhook
>           name: {{ $.Values.encryptedSecrets.name }}
>         text: "{{ range .Alerts }}{{ .Annotations.summary }}{{ .Annotations.description }}{{ .Annotations.message }}{{ end }}"
> ```
> 
> but the text field gets applied as an empty string

Yep, embedding templates in the fields is the work-around I'm also currently using. I eventually got it working with the config below, but it looked like some fields didn't allow multiline strings, specifically the `color` and `action.url` fields. It turned out that using a single-quotted one-line string for those problematic fields allowed it to work.

Here's my workaround AlertmanagerConfig manifest:

```
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: global-alert-manager-config
  labels:
    alertmanagerConfig: slack
  namespace: monitoring
spec:
  route: # root route, all child routes inherit these settings if not overidden.
    groupBy: ["job", "alertname"]
    groupWait: 5s
    groupInterval: 5m
    repeatInterval: 1m # change to 1d for higher environments
    receiver: "slack" # All alerts that do not match the following child routes
    # will remain at the root node and be dispatched to 'default-receiver'.

    routes:
      - receiver: "slack"
        continue: true # Allow alert to continue to go-alert receiver

  receivers:
    - name: "slack" # Slack config templating from https://hodovi.cc/blog/creating-awesome-alertmanager-templates-for-slack/
      slackConfigs:
        - channel: dev-alerts
          apiURL:
            name: alert-manager-slack-webhook
            key: webhookURL
            optional: false
          iconEmoji: "test_tube"
          titleLink: https://www.example.com
          # Specifying the color with multiline YAML string does not work, hence this difficult-to-read single quoted one liner.
          color: '{{ if eq .Status "firing" -}}{{ if eq .CommonLabels.severity "warning" -}}warning{{- else if eq .CommonLabels.severity "critical" -}}danger{{- else -}}#439FE0{{- end -}}{{ else -}}good{{- end }}'
          title: |
            {{/* Title of the Slack alert */}}
            {{ define "slack.title" -}}
            [{{ .Status | toUpper -}}
            {{ if eq .Status "firing" }}:{{ .Alerts.Firing | len }}{{- end -}}
            ] {{ .CommonLabels.alertname }}
            {{- end }}
            {{ template "slack.title" . }}
          text: |
            {{/* Severity of the alert */}}
            {{ define "__alert_severity" -}}
                {{- if eq .CommonLabels.severity "critical" -}}
                *Severity:* `Critical` :red_alert:
                {{- else if eq .CommonLabels.severity "warning" -}}
                *Severity:* `Warning` :warning:
                {{- else if eq .CommonLabels.severity "info" -}}
                *Severity:* `Info` :information_source:
                {{- else -}}
                *Severity:* `Unknown` :question: {{ .CommonLabels.severity }}
                {{- end }}
            {{- end }}
            {{/* The text to display in the alert */}}
            {{ define "slack.text" -}}

                {{ template "__alert_severity" . }}
                {{- if (index .Alerts 0).Annotations.summary }}
                {{- "\n" -}}
                *Summary:* {{ (index .Alerts 0).Annotations.summary }}
                {{- end }}
                {{- if (index .Alerts 0).Labels.namespace }}
                {{- "\n" -}}
                *Namespace:* `{{ (index .Alerts 0).Labels.namespace }}`
                {{- end }}
                {{ range .Alerts }}

                    {{- if .Annotations.description }}
                    {{- "\n" -}}
                    {{ .Annotations.description }}
                    {{- "\n" -}}
                    {{- end }}
                    {{- if .Annotations.message }}
                    {{- "\n" -}}
                    {{ .Annotations.message }}
                    {{- "\n" -}}
                    {{- end }}

                {{- end }}

            {{- end }}
            {{ template "slack.text" . }}
          sendResolved: true
          actions:
            - type: button
              text: "Runbook :green_book:"
              url: "{{ (index .Alerts 0).Annotations.runbook_url }}"
            - type: button
              text: "Query :mag:"
              url: "{{ (index .Alerts 0).GeneratorURL }}"
            - type: button
              text: "Dashboard :chart_with_upwards_trend:"
              url: "{{ (index .Alerts 0).Annotations.dashboard_url }}"
            - type: button
              text: "Silence :no_bell:"
              # Specifying the url with multiline YAML string does not work, hence this difficult-to-read single quoted one liner.
              url: '{{ .ExternalURL }}/#/silences/new?filter=%7B{{- range .CommonLabels.SortedPairs -}}{{- if ne .Name "alertname" -}}{{- .Name }}%3D"{{- .Value -}}"%2C%20{{- end -}}{{- end -}}alertname%3D"{{- .CommonLabels.alertname -}}"%7D'
```

**Comment by DylanWarnock** [comment]
Thank you! Would you mind share your an example of your PrometheusRule CRDs? How are CommonLabels vs Labels defined?

**Comment by slyt** [comment]
> Thank you! Would you mind share your an example of your PrometheusRule CRDs? How are CommonLabels vs Labels defined?

Sure thing, here's the test rule that I used to debug my template formatting, you'll probably have to change the `namespace` fields, but other than that is should apply okay. Note that you don't actually need to define CommonLabels in your alerts; I'm not sure exactly how those work though, [the docs](https://prometheus.io/docs/alerting/latest/notifications/#data) are not very detailed.

```YAML
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    release: prometheus
    role: alert-rules
  name: test-prometheus-rule
  namespace: taylor-sly
spec:
  groups:
    - name: test-group
      rules:
        - alert: test-alert-critical-taylor-sly
          annotations:
            description:
              This is the first test alert in the test-group.
            runbook_url: https://www.example.com
            summary: This is the summary of test-alert-1
          expr: vector(1)
          for: 10s
          labels:
            severity: critical
            namespace: taylor-sly-2
            my_label: test-label-critical
            emoji: ":red_alert:"
        - alert: test-alert-info-taylor-sly
          annotations:
            description:
              This is the first test alert in the test-group.
            runbook_url: https://www.example.com
            summary: This is the summary of test-alert-1
          expr: vector(1)
          for: 10s
          labels:
            severity: info
            namespace: taylor-sly-2
            my_label: test-label-10
            emoji: ":information_source:"
        - alert: test-alert-warning-taylor-sly
          annotations:
            dashboard_url: https://www.example.com
            description:
              The value of the PrometheusRule expression is {{ $value }}.
            runbook_url: https://www.example.com/test-alert-2
            summary: This is the summary and the expression value is {{ $value }}!
          expr: |
            vector(1)
          for: 10s
          labels:
            severity: warning
            source_namespace: "{{ $labels.namespace }}"
            my_label: test-label-20
            emoji: ":warning:"
        - alert: test-alert-default-taylor-sly
          annotations:
            dashboard_url: https://www.example.com
            description:
              The value of the PrometheusRule expression is {{ $value }}.
            runbook_url: https://www.example.com/test-alert-2
            summary: This is the summary and the expression value is {{ $value }}!
          expr: |
            vector(1)
          for: 10s
          labels:
            severity: foobar
            source_namespace: "{{ $labels.namespace }}"
            my_label: test-label-20
            emoji: ":question:"
```

**Comment by DylanWarnock** [comment]
I'm not sure how you are applying your chart. I'm using helm managed by argocd. Have you encountered an issue where helm seems be interpreting the text after `text: |` as template references rather than a multiple string to be passed to alert manager?

**Comment by DylanWarnock** [comment]
Helm keeps saying I'm making reference to values that don't exist, but should the entire block after `text: |` be interpreted as a multiline string?

**Comment by slyt** [comment]
I'm using kube-prometheus-stack and deploying with ArgoCD. However, I try to avoid configuring stuff in the helm values of the ArgoCD Application, it's difficult to patch and syntax highlighting doesn't work well. 

What does you values.yaml look like for your helm chart?

My ArgoCD Application manifest for kube-prometheus-stack looks like this. Note that `alertmanager.alertmanagerSpec.alertmanagerConfiguration.name` points to my AlertManagerConfig I mentioned in my previous comments:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: prometheus
  namespace: argocd
  annotations:
    helm: "true"
spec:
  project: default
  source:
    chart: kube-prometheus-stack
    repoURL: "https://prometheus-community.github.io/helm-charts"
    targetRevision: 41.0.0 # This version upgrades Prometheus-Operator to v0.60.1, Prometheus to v2.39.1, and Thanos to v0.28.1. This version also upgrades the Helm charts of kube-state-metrics to 4.20.2, prometheus-node-exporter to 4.3.0 and Grafana to 6.40.4.
    helm:
      releaseName: prometheus # Match all Prometheus CRDs where labels.release=prometheus
      skipCrds: true
      values: |
        prometheus:
          prometheusSpec:
            namespaceSelector:
              any: 'true'
            remoteWrite:
              - url: 'https://www.example.com/telegraf'
            additionalScrapeConfigsSecret:
              enabled: 'true'
              name: additional-scrape-configs
              key: additional-sources.yaml
        kube-state-metrics: # https://github.com/prometheus-community/helm-charts/blob/6956850383add81f379094a0446fb0a062fd9909/charts/kube-state-metrics/values.yaml#L157-L164
          metricLabelsAllowlist:
            - namespaces=[example-namespace]
        alertmanager:
          alertmanagerSpec:
            alertmanagerConfiguration:
              name: global-alert-manager-config
  destination:
    server: "https://kubernetes.default.svc"
    namespace: monitoring

  syncPolicy:
    syncOptions:
      - CreateNamespace=true

```

**Comment by slyt** [comment]
I've submitted [a seperate issue on the kube-prometheus-stack project](https://github.com/prometheus-community/helm-charts/issues/2657) because I was having issues adding the templates to the values.yaml this way as well.

**Comment by orihomie** [comment]
Also facing the same issue.

**Comment by simonpasquier** [comment]
:wave: we can't really support templates with the AlertmanagerConfig CRD because there's no way to ensure that 2 objects don't create conflicting template definitions. For instance, if AlertmanagerConfig `foo`'s template contains `{{ define "title" }}some nice title{{ end }}` and AlertmanagerConfig `bar`'s template contains `{{ define "title" }}another nice title{{ end }}` then we can't predict which template function is going to be used eventually.

**Comment by slyt** [comment]
> 👋 we can't really support templates with the AlertmanagerConfig CRD because there's no way to ensure that 2 objects don't create conflicting template definitions. For instance, if AlertmanagerConfig `foo`'s template contains `{{ define "title" }}some nice title{{ end }}` and AlertmanagerConfig `bar`'s template contains `{{ define "title" }}another nice title{{ end }}` then we can't predict which template function is going to be used eventually.

Could the templates be scoped so that they only apply to the AlertManagerConfig in the same file? I think that would solved the issues we're all having. Maybe there's no direct support for that type of thing in Prometheus itself though which would make the implementation pretty involved..

**Comment by orihomie** [comment]
> :wave: we can't really support templates with the AlertmanagerConfig CRD because there's no way to ensure that 2 objects don't create conflicting template definitions. For instance, if AlertmanagerConfig `foo`'s template contains `{{ define "title" }}some nice title{{ end }}` and AlertmanagerConfig `bar`'s template contains `{{ define "title" }}another nice title{{ end }}` then we can't predict which template function is going to be used eventually.

It's okay to check for existing templates and raise exception if there's conflicts.

Talking about our subject, it could be the option to pass templates through configmap in a file.

**Comment by simonpasquier** [comment]
I haven't found a way to parse templates in advance and detect conflicts but happy to hear otherwise.

**Comment by orihomie** [comment]
@simonpasquier can you please tell how does default templates are stored now?

**Comment by simonpasquier** [comment]
@orihomie not sure what your question is exactly. You can pass template files to Alertmanager using one of the following methods:
* When using a secret to provide the Alertmanager configuration, additional keys in the secret will be mounted in the /etc/alertmanager/config directory in the alertmanager container  (see https://prometheus-operator.dev/docs/user-guides/alerting/#using-a-kubernetes-secret).
* When using `spec.alertmanagerConfiguration` you can provide secrets & configmaps containing templates (see https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.AlertmanagerConfiguration).
* You can also mount additional secrets & configmaps via `spec.secrets` & `spec.configmaps`.

**Comment by orihomie** [comment]
When I've installed kube-prom-stack, I saw there is already installed this default template in `etc/alertmanager/config/templates.tmpl`:
```
{{ define "slack.myorg.text" }}
{{- $root := . -}}
{{ range .Alerts }}
  *Alert:* {{ .Annotations.summary }} - `{{ .Labels.severity }}`
  *Cluster:* {{ template "cluster" $root }}
  *Description:* {{ .Annotations.description }}
  *Graph:* <{{ .GeneratorURL }}|:chart_with_upwards_trend:>
  *Runbook:* <{{ .Annotations.runbook }}|:spiral_note_pad:>
  *Details:*
    {{ range .Labels.SortedPairs }} - *{{ .Name }}:* `{{ .Value }}`
    {{ end }}
{{ end }}
{{ end }}
```

How did it get there?

I've managed to pass desired templates by upgrading kube-prom-stack release by passing `values.override.yaml`:
```
alertmanager:

  tplConfig: true
  templateFiles:
     templates.tmpl: |-
       ...
```

But that seems complicated to me. It would be much better if all templates would be passed through ConfigMaps, which then will be mounted to a `/etc/alertmanager/config` directory.

Talking about checking for duplicating templates I suppose this is mainly responsibility of the alertmanager application itself to check for conflicting (duplicating) templates, but we can make something like implementing validating admission controller for that purpose, can't we?

**Comment by simonpasquier** [comment]
> How did it get there?

This is a question for the Helm chart maintainers.

> Talking about checking for duplicating templates I suppose this is mainly responsibility of the alertmanager application itself to check for conflicting (duplicating) templates, but we can make something like implementing validating admission controller for that purpose, can't we?

see my previous reply: https://github.com/prometheus-operator/prometheus-operator/issues/5139#issuecomment-1359089609

TL;DR: we haven't found a solution yet.

**Comment by BennyDjo** [comment]
Hello, 

I am a bit confused here on the steps allowing to add new template as a Secret.
 I create a secret containing my templates  named "alertmanager-templates".

```
apiVersion: v1                                                                                                                                                                                                                                                                                                                                              
  kind: Secret                                                                                                                                                                                                                                                                                                                                              
metadata:                                                                                                                                                                                                                                                                                                                                                   
  name: alertmanager-templates                                                                                                                                                                                                                                                                                                                            
data:                                                                                                                                                                                                                                                                                                                                                       
  mytemplate: Q2VjaSBlc3QgdW5lIGFsZXJ0ZQ== `    
```
Where base64 refers to "This is an alert"

 I add the secret in my AlertManagers definition with 
```
alertmanagerConfiguration:                                                                                                                                                                                                                                                                                                                                        
  templates:                                                                                                                                                                                                                                                                                                                                                  
  - secret:                                                                                                                                                                                                                                                                                                                                                     
      key: alertmanager-templates                                                                                                                                                                                                                                                                                                                                 
      name: alertmanager-templates
```

And I try to access the template in my receiver definition in AlertManagerConfig:


```
spec:                                                                                                                                                                                                                                                                                                                                                         
  receivers:                                                                                                                                                                                                                                                                                                                                                
  - name: sns                                                                                                                                                                                                                                                                                                                                          
      snsConfigs:                                                                                                                                                                                                                                                                                                                                                
        message: '{{ template "mytemplate" . }}' 
```                                                                                                                                                                                                                                                                                                     

But I get the following error doing so:

`notify retry canceled due to unrecoverable error after 1 attempts: unexpected status code 400: Invalid parameter: Empty message","level":"error","msg":"Notify for alerts failed`


Is this not the way to be used ?

**Comment by simonpasquier** [comment]
@BennyDjo the `mytemplate` key in the secret needs to hold a valid Go template, e.g. 

````
{{ define "mytemplate" }} This is an alert{{ end}}
````

**Comment by seifrajhi** [comment]
We are facing the exact same issue in our clusters unfortunately

**Comment by nithyasree73** [comment]
Any Updates here ^^, i am currently blocked here where i need to add templates field.

**Comment by nfaust-edb** [comment]
This works for me.

```
kind: AlertManager
spec
  alertmanagerConfiguration
    templates
      - configMap
          name: alertmanager-default-tmpl
          key: default.tmpl
```
---
```
/alertmanager $ cat /etc/alertmanager/config_out/alertmanager.env.yaml
route:
  receiver: monitoring/alertmanager-config/webhook_name
  group_by:
  - alertname
  - cluster
  - service
  routes:
  - receiver: monitoring/alertmanager-config/webhook_name
    continue: true
  group_wait: 30s
  group_interval: 5m
  repeat_interval: 15m
receivers:
- name: monitoring/alertmanager-config/webhook_name
  webhook_configs:
  - url: https://webhook_name.url/api/submit-event?token=****
    max_alerts: 10
templates:
- /etc/alertmanager/templates/default.tmpl

/alertmanager $ ls -alh /etc/alertmanager/templates/
total 4K
drwxrwsrwt    3 root     2000         100 Dec  9 17:47 .
drwxr-xr-x    1 nobody   nobody      4.0K Dec  9 17:47 ..
drwxr-sr-x    2 root     2000          60 Dec  9 17:47 ..2024_12_09_17_47_08.3063119549
lrwxrwxrwx    1 root     2000          32 Dec  9 17:47 ..data -> ..2024_12_09_17_47_08.3063119549
lrwxrwxrwx    1 root     2000          19 Dec  9 17:47 default.tmpl -> ..data/default.tmpl

/alertmanager $ cat /etc/alertmanager/templates/default.tmpl
{{ define "__alertmanager" }}Alertmanager{{ end }}
{{ define "__alertmanagerURL" }}{{ .ExternalURL }}/#/alerts?receiver={{ .Receiver | urlquery }}{{ end }}
```

**Comment by shayak** [comment]
@nfaust-edb 
Does this work with AlertmanagerConfig resources? Could you share the rest of your alertmanager values file?

**Comment by SharifiFaranak** [comment]
was able to get it working, see [here](https://github.com/prometheus-community/helm-charts/issues/3900#issuecomment-3168807655)

@simonpasquier would you have insight into whether `alertmanagerConfiguration` would be going away or not? (since it seems to be in `EXPERIMENTAL` state right now). Would appreciate it not going away :)

**Comment by simonpasquier** [comment]
@SharifiFaranak it's not going to go away. It's marked experimental mostly because it didn't support all existing Alertmanager configuration but I think that we're pretty close now.

**Comment by simonpasquier** [comment]
Closing because we have no way to prevent clashing between template functions. We can only reopen if we find a solution to avoid the issue described in https://github.com/prometheus-operator/prometheus-operator/issues/5139#issuecomment-1358006335
